# Exercise 1
Import `US_ACS_2017_10pct_sample.dta` into a pandas DataFrame (read it directly from a URL to help the autograder, please).

This can be done with the command `pd.read_stata`, which will read in files created in the program Stata (and which uses the file suffix `.dta`). This is a format commonly used by social scientists.

In [1]:
import pandas as pd

df = pd.read_stata(
    "https://github.com/nickeubank/MIDS_Data/raw/master/US_AmericanCommunitySurvey/US_ACS_2017_10pct_sample.dta"
)

# Exercise 2
How many observations are in your data? Store the answer in your results dictionary with the key "ex2_num_obs".

In [2]:
ex2_num_obs = df.shape[0]
results = {"ex2_num_obs": ex2_num_obs}

# Exercise 3
How many variables are in your data? Store the answer in your results dictionary with the key "ex3_num_vars".

In [3]:
ex3_num_vars = df.shape[1]
results = {"ex3_num_vars": ex3_num_vars}

# Exercise 4

In [4]:
df.columns

Index(['year', 'datanum', 'serial', 'cbserial', 'numprec', 'subsamp', 'hhwt',
       'hhtype', 'cluster', 'adjust',
       ...
       'migcounty1', 'migmet131', 'vetdisab', 'diffrem', 'diffphys', 'diffmob',
       'diffcare', 'diffsens', 'diffeye', 'diffhear'],
      dtype='object', length=104)

In [5]:
for c in df.columns:
    print(c)

year
datanum
serial
cbserial
numprec
subsamp
hhwt
hhtype
cluster
adjust
cpi99
region
stateicp
statefip
countyicp
countyfip
metro
city
citypop
strata
gq
farm
ownershp
ownershpd
mortgage
mortgag2
mortamt1
mortamt2
respmode
pernum
cbpernum
perwt
slwt
famunit
sex
age
marst
birthyr
race
raced
hispan
hispand
bpl
bpld
citizen
yrnatur
yrimmig
language
languaged
speakeng
hcovany
hcovpriv
hinsemp
hinspur
hinstri
hcovpub
hinscaid
hinscare
hinsva
hinsihs
school
educ
educd
gradeatt
gradeattd
schltype
degfield
degfieldd
degfield2
degfield2d
empstat
empstatd
labforce
occ
ind
classwkr
classwkrd
looking
availble
inctot
ftotinc
incwage
incbus00
incss
incwelfr
incinvst
incretir
incsupp
incother
incearn
poverty
migrate1
migrate1d
migplac1
migcounty1
migmet131
vetdisab
diffrem
diffphys
diffmob
diffcare
diffsens
diffeye
diffhear


# Exercise 5
Subset your data to just those variables: `age, inctot, educ, and empstat`

In [6]:
# List of variables of interest
vars_of_interest = ["age", "inctot", "educ", "empstat"]

# Subset the DataFrame to include only the specified variables
subset_df = df[vars_of_interest]

# Exercise 6
Random sample of 20 rows.

In [7]:
random_sample = subset_df.sample(n=20)

# Exercise 7
Do you see any immediate problems? What issues do you see?

>There are some $9,999,999 as income in the dataset

# Exercise 8
Dropping anyone who has `inctot` equal to `9999999`.

After dropping, how many observations do you have? Save your answer in your results dictionary under the key "ex8_updated_num_obs"

In [8]:
subset_df_drop_9 = subset_df[subset_df["inctot"] != 9999999]
ex8_updated_num_obs = subset_df_drop_9.shape[0]
results = {"ex8_updated_num_obs": ex8_updated_num_obs}

# Exercise 9
Subset the dataset for the people for whom empstat is equal to “employed”. 
Save your answer in your results dictionary under the key "ex9_updated_num_obs"

In [9]:
subset_df_employed = subset_df_drop_9[subset_df_drop_9["empstat"] == "employed"]
subset_df_employed1 = subset_df[subset_df["empstat"] == "employed"]
ex9_updated_num_obs = subset_df_employed.shape[0]
results = {"ex9_updated_num_obs": ex9_updated_num_obs}

# Exercise 10

In [10]:
df["educ"].value_counts()

educ
grade 12                     93133
4 years of college           47212
1 year of college            38779
5+ years of college          29801
nursery school to grade 4    24514
grade 5, 6, 7, or 8          21535
2 years of college           20757
n/a or no schooling          19562
grade 11                      8758
grade 10                      7818
grade 9                       7135
Name: count, dtype: int64

# Exercise 11
There are a lot of values in here, so let’s just check a couple. What is the average value of `inctot` for people whose highest grade level is “grade 12” (in the US, that is someone who has graduated high school)?

Save your answer in your results dictionary under the key "ex11_grade12_income".

In [11]:
grade_12_df = subset_df_employed[subset_df_employed["educ"] == "grade 12"]
average_inc_grade_12 = grade_12_df["inctot"].mean()
results = {"ex11_grade12_income": average_inc_grade_12}

# Exercise 12
What is the average income of someone who has completed an undergraduate degree but not done any postgraduate education (“4 years of college”)?

Save your answer in your results dictionary under the key "ex12_college_income".

In percentage terms, how much does an employed college graduate earn as compared to someone who is only a high school graduate? Use the reference category that gives an answer above 100.

Store your answer in "ex12_college_income_pct". Put your answer in percentage terms (so 100 implies they earn the same amount).

Make sure to interpret your result in words when you print it out!

In [12]:
undergrad_df = subset_df_employed[subset_df_employed["educ"] == "4 years of college"]
average_income_undergrad = undergrad_df["inctot"].mean()

percent = (average_income_undergrad / average_inc_grade_12) * 100
results = {
    "ex12_college_income": average_income_undergrad,
    "ex12_college_income_pct": percent,
}
percent

193.7612727527617

> Employed college graduates earn 193.76% more than those who are only high school graduates.

# Exercise 13
What does that suggest is the value of getting a college degree after graduating high school?

> Getting a college degree has a higher chance of getting a better pay.

# Exercise 14
What is the average income for someone who has not finished high school? What does that suggest is the value of a high school diploma? (Treat n/a or no schooling as having no formal schooling, not as missing).

Hint: You may find the .isin() method to be really helpful here.

Save your answer in your results dictionary under the key "ex14_high_school_dropout".

In [13]:
# Define a list that they do not finish high school
not_finished = [
    "grade 9",
    "grade 10",
    "grade 11",
    "n/a or no schooling",
    "nursery school to grade 4",
    "grade 5, 6, 7, or 8",
]
# Filter the data frame for those who do not finish high school
not_finished_high_school_df = subset_df_employed[
    subset_df_employed["educ"].isin(not_finished)
]

average_income_no_finished = not_finished_high_school_df["inctot"].mean()
results = {"ex14_high_school_dropout": average_income_no_finished}

# Exercise 15
Complete the following table (storing values under the provided keys where listed):

Average income for someone who only completed 9th grade (ex15_grade_9): _________

Average income for someone who only completed 10th grade (ex15_grade_10): _________

Average income for someone who only completed 11th grade (ex15_grade_11): _________

Average income for someone who finished high school (12th grade) but never started college (ex15_grade_12): _________

Average income for someone who completed 4 year of college (in the US, this corresponds to getting an undergraduate degree), but has no post-graduate education (no more than 4 years, ex15_4_years_of_college): _________

Average income for someone who has some graduate education (more than 4 years, ex15_graduate): _________

In [14]:
grade_9_df = subset_df_employed[subset_df_employed["educ"] == "grade 9"]
average_grade_9 = grade_9_df["inctot"].mean()
grade_10_df = subset_df_employed[subset_df_employed["educ"] == "grade 10"]
average_grade_10 = grade_10_df["inctot"].mean()
grade_11_df = subset_df_employed[subset_df_employed["educ"] == "grade 11"]
average_grade_11 = grade_11_df["inctot"].mean()
graduate_df = subset_df_employed[subset_df_employed["educ"] == "5+ years of college"]
average_graduate = graduate_df["inctot"].mean()
results = {
    "ex15_grade_9": average_grade_9,
    "ex15_grade_10": average_grade_10,
    "ex15_grade_11": average_grade_11,
    "ex15_grade_12": average_inc_grade_12,
    "ex15_4_years_of_college": average_income_undergrad,
    "ex15_graduate": average_graduate,
}

# Exercise 16
Why do you think there is no benefit from moving from grade 9 to grade 10, or grade 10 to grade 11, but there is a huge benefit to moving from grade 11 to graduating high school (grade 12)?

> Finishing high school means you reach a milestone of finishing 12 years of studying, hence proving you do not give up easily.

In [15]:
results = {
    "ex2_num_obs": ex2_num_obs,
    "ex3_num_vars": ex3_num_vars,
    "ex8_updated_num_obs": ex8_updated_num_obs,
    "ex9_updated_num_obs": ex9_updated_num_obs,
    "ex11_grade12_income": average_inc_grade_12,
    "ex12_college_income": average_income_undergrad,
    "ex12_college_income_pct": percent,
    "ex14_high_school_dropout": average_income_no_finished,
    "ex15_grade_9": average_grade_9,
    "ex15_grade_10": average_grade_10,
    "ex15_grade_11": average_grade_11,
    "ex15_grade_12": average_inc_grade_12,
    "ex15_4_years_of_college": average_income_undergrad,
    "ex15_graduate": average_graduate,
}

In [17]:
results

{'ex2_num_obs': 319004,
 'ex3_num_vars': 104,
 'ex8_updated_num_obs': 265103,
 'ex9_updated_num_obs': 148758,
 'ex11_grade12_income': 38957.76068179442,
 'ex12_college_income': 75485.05293301983,
 'ex12_college_income_pct': 193.7612727527617,
 'ex14_high_school_dropout': 26226.45692998571,
 'ex15_grade_9': 27171.907751937986,
 'ex15_grade_10': 23018.795811518325,
 'ex15_grade_11': 21541.68693119767,
 'ex15_grade_12': 38957.76068179442,
 'ex15_4_years_of_college': 75485.05293301983,
 'ex15_graduate': 110013.2213384139}

In [16]:
assert set(results.keys()) == {
    "ex2_num_obs",
    "ex3_num_vars",
    "ex8_updated_num_obs",
    "ex9_updated_num_obs",
    "ex11_grade12_income",
    "ex12_college_income",
    "ex12_college_income_pct",
    "ex14_high_school_dropout",
    "ex15_grade_9",
    "ex15_grade_10",
    "ex15_grade_11",
    "ex15_grade_12",
    "ex15_4_years_of_college",
    "ex15_graduate",
}